### PR TITLE
F-123 Add raw section

### LINKED ISSUE #123: [Template] Add support for RAW
I would like create a template that starts an UEFI image.
I need, in the template, something like:
```
RAW = [
  DATA = "
  <os firmware='efi'>
    <type>hvm</type>
    <loader secure='no'/>
  </os>
",
  TYPE = "kvm" ]
```

### PR DESCRIPTION
**Pull request by treywelsh**

### CODE REVIEW

File: opennebula/resource_opennebula_template.go [hunk 1]
```diff
@@ -48,6 +48,36 @@ func resourceOpennebulaTemplate() *schema.Resource {
...
+			"raw": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				MaxItems:    1,
+				Description: "Low-level hypervisor tuning",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (jaypif) [inline opennebula/resource_opennebula_template.go]: why optional ? It is mandatory no ?

File: opennebula/resource_opennebula_template.go [hunk 2]
```diff
@@ -48,6 +48,36 @@ func resourceOpennebulaTemplate() *schema.Resource {
...
+									errors = append(errors, fmt.Errorf("Type %q must be one of: %s", k, strings.Join(validtypes, ",")))
+								}
+
+								return
+							},
+							Description: "Name of the hypervisor: kvm, lxd, vmware",
+						},
+						"data": {
+							Type:        schema.TypeString,
+							Optional:    true,
```
Comment (jaypif) [inline opennebula/resource_opennebula_template.go]: it is also mandatory to me

File: opennebula/resource_opennebula_template.go [hunk 3]
```diff
@@ -272,11 +302,46 @@ func resourceOpennebulaTemplateRead(d *schema.ResourceData, meta interface{}) er
...
+
+		if _, ok := d.GetOk("raw"); ok {
+			err = d.Set("raw", rawMap)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if tagsInterface, ok := d.GetOk("tags"); ok {
```
Comment (jaypif) [inline opennebula/resource_opennebula_template.go]: not related to the issue you are fixing

File: opennebula/resource_opennebula_virtual_machine.go [hunk 4]
```diff
@@ -492,14 +492,23 @@ func resourceOpennebulaVirtualMachineReadCustom(d *schema.ResourceData, meta int
...
-	err = flattenTemplate(d, &vm.Template, false)
+	err = flattenTemplate(d, &vm.Template)
 	if err != nil {
 		return err
 	}
 
-	err = flattenTags(d, &vm.UserTemplate)
-	if err != nil {
-		return err
+	if tagsInterface, ok := d.GetOk("tags"); ok {
```
Comment (jaypif) [inline opennebula/resource_opennebula_virtual_machine.go]: not related to the issue you are fixing

File: opennebula/shared_schemas.go [hunk 5]
```diff
@@ -446,7 +447,22 @@ func flattenDisk(disk shared.Disk) map[string]interface{} {
 	}
 }
 
-func flattenTemplate(d *schema.ResourceData, vmTemplate *vm.Template, tplTags bool) error {
+func flattenTags(tagsConfig map[string]interface{}, template *dyn.Template) (map[string]interface{}, error) {
```
Comment (jaypif) [inline opennebula/shared_schemas.go]: not related to the issue you are fixing

**Review by jaypif — CHANGES_REQUESTED** [review]

File: opennebula/resource_opennebula_template.go (hunk 1 above)
Comment (treywelsh) [reply opennebula/resource_opennebula_template.go]: I probably misunderstood the doc at first (https://docs.opennebula.io/5.12/operation/references/template.html#hypervisor-section), in sunstone it seems required

**Review by jaypif — APPROVED** [review]

**Pull request merged**